firedrake/slate/slac/utils.py

```diff
@@ -176,15 +177,6 @@ def _slate2gem_tensor(expr, self):
     return var
 
 
-@_slate2gem.register(sl.Block)
-def _slate2gem_block(expr, self):
-    child, = map(self, expr.children)
-    child_shapes = expr.children[0].shapes
-    offsets = tuple(sum(shape[:idx]) for shape, (idx, *_)
-                    in zip(child_shapes.values(), expr._indices))
-    return view(child, *(slice(idx, idx+extent) for idx, extent in zip(offsets, expr.shape)))
```

Comment:
I wonder if, rather than removing this functionality of indexing into general expressions, we should have a preprocessing pass on slate that pushes indexing as far towards terminals as possible. e.g. `Block(Transpose(A), (0, 1)) == Transpose(Block(A, (1, 0))) == Transpose(extract_sub_block(A, (1, 0))`. Then you get the nice small blocks that you need, but if someone writes some complex expression and then takes some blocks from it, it all still works.

WDYT?

Reply:
How are we going to do that for matrix-matrix multiplication?

Reply:
I haven't thought it through, but algebraically it is possible, e.g.:
```
[A, B [E, F   = [AE + BG, AF + BH
 C, D] G, H]     CE + DG, CF + DH]

([A, B [E, F       = ([AE + BG, AF + BH
 C, D] G, H])[0,0]     CE + DG, CF + DH])[0, 0] = AE + BG
```

Reply:
Okay so for this example that would be
```
Block(Mul(M1,M2), (0,0)) = Add(Add(Block(M1, (0,0)),  Block(M2, (0,0))),
                               Add(Block(M1, (0,1)),  Block(M2, (1,0)))
```

What about inverse and solve?

Reply:
For inverse you can, if you want, do it with a schur complement factorisation, e.g. (unchecked)

```
[A, B ^{-1} = [I ,-A^{-1} B  [A^{-1} 0              [I 0 
 C D]          0 , I]         0     (D - CA^{-1} B)^{-1}] -CA^{-1} I]

= [A^{-1} + A^{-1} B (D - CA^{-1} B)^{-1} C A^{-1}, -A^{-1} B (D - C A^{-1} B)^{-1}
   -(D - CA^{-1} B)^{-1} CA^{-1},  D - C^{-1} B^{-1} ]
```

But this is fiddly to get right because to do the elimination this way requires that A is invertible (you could also do it the other way, requiring that D is invertible) see the "Derivation from LDU decomposition" section at https://en.wikipedia.org/wiki/Woodbury_matrix_identity.

So for inverse/solve I would regard those as "stopping points" and not try and push the indexing inside.

Reply:
Okay I will go for the latter option for now and then we can redecide later if needed.